+nosnoc/+discrete_time_problem/Gcs.m

```diff
@@ -0,0 +1,870 @@
+        end
+
+        function create_solver(obj, solver_options, plugin)
+            if ~obj.populated
+                obj.populate_problem();
+            end
+
+            if ~exist('plugin')
+                plugin = 'scholtes_ineq';
+            end
+
+            % Sort by indices to recover almost block-band structure.
+            obj.w.sort_by_index();
+            obj.g.sort_by_index();
+
+            solver_options.assume_lower_bounds = true;
+
+            obj.solver = nosnoc.solver.mpccsol('Mpcc solver', plugin, obj.to_casadi_struct(), solver_options);
+        end
+
+        function stats = solve(obj)
+            opts = obj.opts;
+            T_val = obj.p.T().val;
+
+            for ii=1:opts.N_stages
+                for jj=1:opts.N_finite_elements(ii)
+                    % Recalculate ubh and lbh based on T_val
+                    h0 = T_val/(opts.N_stages*opts.N_finite_elements(ii));
+                    ubh = (1 + opts.gamma_h) * h0;
+                    lbh = (1 - opts.gamma_h) * h0;
+                    if opts.time_rescaling && ~opts.use_speed_of_time_variables
+                        % if only time_rescaling is true, speed of time and step size all lumped together, e.g., \hat{h}_{k,i} = s_n * h_{k,i}, hence the bounds need to be extended.
+                        ubh = (1+opts.gamma_h)*h0*opts.s_sot_max;
+                        lbh = (1-opts.gamma_h)*h0/opts.s_sot_min;
+                    end
+                    obj.w.h(ii,jj).lb = lbh;
+                    obj.w.h(ii,jj).ub = ubh;
+                end
+            end
+
+            stats = solve@vdx.problems.Mpcc(obj);
+        end
+
+        function results = get_results_struct(obj)
+            opts = obj.opts;
+            model = obj.model;
+
+            rbp = ~opts.right_boundary_point_explicit;
+            
+            if opts.right_boundary_point_explicit
+                results.x = obj.discrete_time_problem.w.x(:,:,obj.opts.n_s).res;
+                results.z = obj.discrete_time_problem.w.z(:,:,obj.opts.n_s).res;
+                results.lambda = obj.discrete_time_problem.w.lambda(:,:,obj.opts.n_s).res;
+                results.mu = obj.discrete_time_problem.w.mu(:,:,obj.opts.n_s).res;
+                results.theta = obj.discrete_time_problem.w.theta(:,:,obj.opts.n_s).res;
+            else
+                results.x = [obj.discrete_time_problem.w.x(0,0,obj.opts.n_s).res,...
+                    obj.discrete_time_problem.w.x(1:opts.N_stages,:,obj.opts.n_s+1).res];
+                results.z = [obj.discrete_time_problem.w.z(0,0,obj.opts.n_s).res,...
+                    obj.discrete_time_problem.w.z(1:opts.N_stages,:,obj.opts.n_s+1).res];
+                results.lambda = [obj.discrete_time_problem.w.lambda(0,0,obj.opts.n_s).res,...
+                    obj.discrete_time_problem.w.lambda(1:opts.N_stages,:,obj.opts.n_s+1).res];
+                results.mu = [obj.discrete_time_problem.w.mu(0,0,obj.opts.n_s).res,...
```

Comment:
i am confused, why do you have a mu and theta in a GCS? they remind me of stewart dcs...
maybe pick different names + explain them

Reply:
This is a copy pate error :sweat_smile:. removed.